components/script/script_thread.rs

```diff
@@ -1471,28 +1476,292 @@ impl ScriptThread {
+                    let touch_result = self.handle_touch_event(
+                        pipeline_id,
+                        event_type,
+                        identifier,
+                        point,
+                        node_address,
+                    );
+                    match (event_type, touch_result) {
+                        (TouchEventType::Down, TouchEventResult::Processed(handled)) => {
+                            let result = if handled {
+                                // TODO: Wait to see if preventDefault is called on the first touchmove event.
+                                EventResult::DefaultAllowed
+                            } else {
+                                EventResult::DefaultPrevented
+                            };
+                            let message = ScriptMsg::TouchEventProcessed(result);
+                            self.script_sender.send((pipeline_id, message)).unwrap();
+                        },
+                        _ => {
+                            // TODO: Calling preventDefault on a touchup event should prevent clicks.
+                        },
+                    }
+                },
+
+                CompositorEvent::WheelEvent(delta, point, node_address) => {
+                    self.handle_wheel_event(pipeline_id, delta, point, node_address);
+                },
+
+                CompositorEvent::KeyboardEvent(key_event) => {
+                    document.dispatch_key_event(key_event);
+                },
+
+                CompositorEvent::IMEDismissedEvent => {
+                    document.ime_dismissed();
+                },
+
+                CompositorEvent::CompositionEvent(composition_event) => {
+                    document.dispatch_composition_event(composition_event);
+                },
+
+                CompositorEvent::GamepadEvent(gamepad_event) => {
+                    let global = window.upcast::<GlobalScope>();
+                    global.handle_gamepad_event(gamepad_event);
+                },
+            }
+        }
+        ScriptThread::set_user_interacting(false);
+    }
+
+    /// <https://html.spec.whatwg.org/multipage/#update-the-rendering>
+    fn update_the_rendering(&self) {
+        // reset batching.
+        *self.has_queued_update_the_rendering_task.borrow_mut() = false;
 
-        // Handle pending resize events.
-        // Gather them first to avoid a double mut borrow on self.
-        let mut resizes = vec![];
+        if self.can_continue_running_inner() {
+            // TODO: Filter non-renderable documents.
+            // TODO: Unnecessary rendering.
+            let pipeline_and_docs: Vec<(PipelineId, DomRoot<Document>)> = self
+                .documents
+                .borrow()
+                .iter()
+                .map(|(id, document)| (id, DomRoot::from_ref(&*document)))
+                .collect();
+            for (pipeline_id, document) in pipeline_and_docs {
```

Comment:
This runs all steps for for every document, but the specification says has "for *doc* in *docs*" for every step. It might be worth putting a comment here that each step needs to iterate through the list of documents.